### ISSUE TITLE
Not all tests run in parallel execution. DotNet Core 2.0

### ISSUE THREAD

**Issue by funtik** [nunit/nunit#2766]
I run 12 test in parallel, but were done only 1 or 2 tests.

I have this `
[assembly: Parallelizable(ParallelScope.All)]
[assembly: LevelOfParallelism(3)]` in my AssemblyInfo.cs

> Starting test execution, please wait...
> NUnit Adapter 3.10.0.21: Test execution started
> Running all tests in f:\Repos\JS_Framework\e2e\bin\Debug\netcoreapp2.0\LuxiteTests.dll
> NUnit3TestExecutor converted 12 of 12 NUnit test cases
> Skipped  VerifyFirstSample
> Error Message:
>  Because it's sample
> 
> The active test run was aborted. Reason: Unhandled Exception: OpenQA.Selenium.NotFoundException: Context was not initialized for this test!!!
>    at LuxiteTests.__Support.Implementation.SessionWithVideoRecording.CaptureSrceenshotForVideo() in f:\Repos\JS_Framework\e2e\__Support\Implementation\SessionWithVideoRecording.cs:line 74
>    at System.Threading.Thread.ThreadMain_ThreadStart()
>    at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
> 
> 
> Total tests: Unknown. Passed: 1. Failed: 0. Skipped: 1.
> Test Run Aborted.
> Test execution time: 3,6764 Minutes

**Comment by jnm2** [comment]
It looks like your test is causing a background thread to throw an exception which terminates the VSTest engine process. Several things can cause this. Is your code open source or do you have a minimal reproduction to share?

**Comment by rprouse** [comment]
@funtik are your tests thread safe? It looks like you are sharing the Selenium Context between tests. That would likely cause it to get setup and torn down by each test, so another test using it would crash. Your test code needs to be thread safe just like any other code you write. That is why it isn't enabled by default.

**Comment by EllaVader** [comment]
@rprouse could you elaborate on what you mean by that?  I think I may be seeing the same type of behavior. Right now I have this defined in my assembly info file:
`[assembly: Parallelizable(ParallelScope.Fixtures)]`
Which works great for parallel runs at the class level, but if  I set ParallelScope to `.children` or `.self` (so I can run ALL of my selenium test methods in parallel) it just isn't working.  

What happens is (for example if i am running 2 selenium tests in parallel at the method level) two browsers open one is blank the other isn't and then all the tests being run are trying to operate on the one browser window.  And finally, that browser window that is blank never closes. 

Could it be that I'm having a problem with threading?  If it is, what would I need to do to fix it?  Or my setup and tear downs?  I also have a onetimesetup method as well.  I would love to be able to get this to work.

**Comment by jnm2** [comment]
Ah, this is probably because people expect us to be doing [instance-per-test-case](https://github.com/nunit/nunit/issues/2574). Right now parallelism more granular than the fixture level cannot be used if you have any setup or teardown methods since the same fixture instance is shared by all test cases.

**Comment by rprouse** [comment]
@EllaVader you are definitely having a problem with threading. NUnit creates one copy of your test class and runs all `SetUp -> Test -> TearDown` in parallel, so if any of those are creating, destroying, or possibly even accessing the same member variables in your class, then your tests are going to stomp on each other.

In your case, two setup methods run in parallel creating the two browser windows, but you are using one variable to control both, so it gets overwritten by the second call. All of your tests then run in the one browser window because the reference to the second was lost.

The easiest way to fix this is to only use local variables in your test methods and do the setup and teardown in those methods.

As @jnm2 states, there are plans to support scenarios like yours more safely by instantiating a copy of the test class for each test. That is expensive and slows down performance, so it would not be the default.

**Comment by garora** [comment]
@rprouse - Tests are not working for .NET Core 2.0. There is no option to down-grade my project to .NET Core 1.1 - [TDD-Katas](https://github.com/garora/TDD-Katas)
Any plan to resolve the issues?

**Comment by jnm2** [comment]
If you issue is because NUnit started parallelizing, the workaround for now is simply to put `[NonParallelizable] on whichever fixtures or methods (or assemblies) that are not thread-safe. Or, do everything inside the test method so that there is no `[SetUp]` or `[TearDown]` in the fixture.

Those are better options than downgrading. Eventually, if enough people comment on https://github.com/nunit/nunit/issues/2574 (or offer to implement it), the instance-per-test-case feature will become available.

**Comment by garora** [comment]
@jnm2 - this is not a case of `parallelizing` etc. Actually, there were build errors in `Visual Studio 2017`
So, we downgraded the project and now we have two different flavors of the same project:
1. [.NET Core 2.0 flavor](https://github.com/garora/TDD-Katas) - _Not working_
2. [.NET Core 1.1 flavor](https://github.com/garora/TDD-Katas-NETCore) - _working perfectly_

**Comment by jnm2** [comment]
@garora If your issue is not about parallelization, then it is not related to this thread. We are very interested in hearing about it though. Please search for the build error message in our GitHub issues and then open a new one if you don't see it. (Just a guess- is it https://github.com/nunit/nunit/issues/2760?)

**Comment by mikkelbu** [comment]
@garora I did not get any build errors? But I was able to make the tests run (to begin with I got an "The following TestContainer was not found" error from VS) by making the following changes in `TDD-Katas-NETCore.csproj`:
```diff
-    <PackageReference Include="NUnit" Version="3.6.1" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.3.0" />
+    <PackageReference Include="NUnit" Version="3.10.1" />
```

Ps. The test `AddReturnSumWhenSuppliedMultipleNumbersInString("0,3,2",5)'` is duplicated.

**Comment by garora** [comment]
@mikkelbu thanks for spending time to help me to resolve this issue. Please note that

1. [.NET Core 2.0 flavor](https://github.com/garora/TDD-Katas) - _Not working_
2. [.NET Core 1.1 flavor](https://github.com/garora/TDD-Katas-NETCore) - _working perfectly_

It looks like you tried on .NET Core 1.1 repository. .NET Core 2.0 is having the same issue - I resolved the build issues but tests are not running.

Thanks for notifying duplicated test, good catch!

**Comment by mikkelbu** [comment]
@garora No I did try the project `TDD-Katas` (first item in the list), and if I make the changes I mentioned above then I can run the tests in VS'17 (note that I've unloaded the framework project to avoid confusion with multiple tests with the same name).
![image](https://user-images.githubusercontent.com/1007631/37991374-4944b2ea-3209-11e8-99c1-977ba027ceca.png)

**Comment by garora** [comment]
@mikkelbu I can see in the snapshot, there are conflict ions in Nuget packages and the same I am facing right now. Its not working at my end. If you are able to run the tests, could you please make a `PR` to `develop` branch?

**Comment by mikkelbu** [comment]
@garora PR supplied here garora/TDD-Katas#40, but I think we are very off-topic wrt. "Not all tests run in parallel execution. DotNet Core 2.0" 😄

**Comment by garora** [comment]
@mikkelbu Thanks for your support. I agree, this issue is not for this discussion: `Make VS discovers tests`. Do you want me to create a new issue to help others so that in future someone need something similar?

**Comment by mikkelbu** [comment]
@garora I don't think it is necessary. The information is already on the wiki: https://github.com/nunit/docs/wiki/.NET-Core-and-.NET-Standard, but if you think anything is missing, then please create an issue in the docs project.
